arviz/stats/stats_utils.py

```diff
@@ -0,0 +1,334 @@
+
+    b_inv overwrites b unless b_inv is None.
+    """
+    # check dimensions for result arrays
+    ary = np.asarray(ary)
+    if ary.dtype.kind == "i":
+        ary = ary.astype(np.float64)
+    dtype = ary.dtype.type
+    shape = ary.shape
+    shape_len = len(shape)
+    if isinstance(axis, Sequence):
+        axis = tuple(axis_i if axis_i >= 0 else shape_len + axis_i for axis_i in axis)
+        agroup = axis
+    else:
+        axis = axis if (axis is None) or (axis >= 0) else shape_len + axis
+        agroup = (axis,)
+    shape_max = (
+        tuple(1 for _ in shape)
+        if axis is None
+        else tuple(1 if i in agroup else d for i, d in enumerate(shape))
+    )
+    # create result arrays
+    if out is None:
+        if not keepdims:
+            out_shape = (
+                tuple()
+                if axis is None
+                else tuple(d for i, d in enumerate(shape) if i not in agroup)
+            )
+        else:
+            out_shape = shape_max
+        out = np.empty(out_shape, dtype=dtype)
+    if b_inv == 0:
+        return np.full_like(out, np.inf, dtype=dtype) if out.shape else np.inf
+    if b_inv is None and b == 0:
+        return np.full_like(out, -np.inf) if out.shape else -np.inf
+    ary_max = np.empty(shape_max, dtype=dtype)
+    # calculations
+    ary.max(axis=axis, keepdims=True, out=ary_max)
+    if copy:
+        ary = ary.copy()
+    ary -= ary_max
+    np.exp(ary, out=ary)
+    ary.sum(axis=axis, keepdims=keepdims, out=out)
+    np.log(out, out=out)
+    if b_inv is not None:
+        ary_max -= np.log(b_inv)
+    elif b:
+        ary_max += np.log(b)
+    out += ary_max.squeeze() if not keepdims else ary_max
+    # transform to scalar if possible
+    return out if out.shape else dtype(out)
+
+
+def _rint(num):
+    """Round and change to ingeter."""
+    rnum = np.rint(num)  # pylint: disable=assignment-from-no-return
+    return int(rnum)
+
+
+def _round(num, decimals):
+    """Skip rounding if decimals is None."""
+    if decimals is not None:
+        num = np.round(num, decimals)
+    return num
+
+
+def _quantile(ary, quantile, axis=None, limit=None):
+    """Use same quantile function as R (Type 7)."""
+    if limit is None:
+        limit = tuple()
+    return mquantiles(ary, quantile, alphap=1, betap=1, axis=axis, limit=limit)
+
+
+def check_nan(ary, axis=None, how="any"):
+    """Check if ary has NaN values."""
+    isnan = np.isnan(ary)
+    if how.lower() == "any":
+        isnan = isnan.any(axis)
+    elif how.lower() == "all":
+        isnan = isnan.all(axis)
+    return isnan
+
+
+def check_valid_size(ary, msg, n=2):
+    """Validate 2D array shape."""
+    ary = np.asarray(ary)
+    shape = ary.shape
+    if len(shape) != 2:
+        raise TypeError("{} calculation requires 2 dimensional array.".format(msg))
+    n_chain, n_draw = shape
+    if n_chain < n:
+        raise TypeError("{} calculation requires multiple chains.".format(msg))

```

Comment:
I would include the n in this message too, like in the draws error message. This is the error risen in the tests currently failing for forest plots with `ess=True`. The data used for testing has only 2 chains, and ess checks for <3, thus, this error is risen even though there are already multiple chains present.

Reply:
I think this should have been fixed. If not, clean these in other PR.